### PR TITLE
Add Whisper inference with quantization notebook

### PR DESCRIPTION
**Pull request by jimypbr**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: notebooks/whisper-quantized-example.ipynb [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eff8bcf6",
+   "metadata": {},
+   "source": [
+    "You can also use a custom checkpoint from Hugging Face Hub using the argument `custom_checkpoint` above. In this case, you have to make sure that `size` matches the checkpoint model size.\n",
+    "\n",
+    "Two features of Optimum Graphcore are demonstrated below:\n",
+    "1. `use_cond_encoder` : This enables putting the Whisper encoder and decoder on a single IPU and switching between them using a compiled `cond` operation. This is only available if `ipus_per_replica == 1`.\n",
```
Comment (katalinic-gc) [inline notebooks/whisper-quantized-example.ipynb:239]: Executable caching does not work for me when this is enabled in a notebook. Did you try it with the feature and checked it works for you?

Comment (jimypbr) [reply notebooks/whisper-quantized-example.ipynb:239]: I just ran it again and it indeed loaded from cache. Using 3.3 SDK.

File: notebooks/whisper-quantized-example.ipynb [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fde99b10-e2d2-4787-877f-fb120e327ccb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Install optimum from source \n",
+    "!pip install git+https://github.com/huggingface/optimum-graphcore.git"
```
Comment (katalinic-gc) [inline notebooks/whisper-quantized-example.ipynb:89]: Technically we could do 0.7.1?

File: notebooks/whisper-quantized-example.ipynb [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    "        IPUConfig(executable_cache_dir=cache_dir,\n",
+    "                  ipus_per_replica=1,\n",
+    "                  explicit_ir_inference=True,\n",
+    "                 )),\n",
+    "\n",
+    "    \"small\": (\"openai/whisper-small.en\", \n",
+    "        IPUConfig(executable_cache_dir=cache_dir,\n",
+    "                  ipus_per_replica=1,\n",
+    "                  matmul_proportion=0.1,\n",
+    "                  explicit_ir_inference=True,\n",
```
Comment (katalinic-gc) [inline notebooks/whisper-quantized-example.ipynb:182]: Tiny, small, and base should already fit on one IPU comfortably when num beams is set to 1, why is explicit IR enabled?

Are the 0.1 matmul proportion settings necessary? I think the less settings the better.

File: notebooks/whisper-quantized-example.ipynb [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f919498b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_checkpoint, ipu_config = select_whisper_config(\"small\", custom_checkpoint=None) "
```
Comment (katalinic-gc) [inline notebooks/whisper-quantized-example.ipynb:228]: Some of the text makes it sound like tiny is being run in the notebook, but small is being chosen.

File: notebooks/whisper-quantized-example.ipynb [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+    "        IPUConfig(executable_cache_dir=cache_dir,\n",
+    "                  ipus_per_replica=4,\n",
+    "                  matmul_proportion=0.1,\n",
+    "                  inference_projection_serialization_factor=5,\n",
+    "                  explicit_ir_inference=True,\n",
+    "                 )),\n",
+    "\n",
+    "    \"large\": (\"openai/whisper-large-v2\", \n",
+    "        IPUConfig(executable_cache_dir=cache_dir,\n",
+    "                  ipus_per_replica=8,\n",
```
Comment (katalinic-gc) [inline notebooks/whisper-quantized-example.ipynb:195]: Remove and uncomment line below?

File: notebooks/whisper-quantized-example.ipynb [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    "    \"small\": (\"openai/whisper-small.en\", \n",
+    "        IPUConfig(executable_cache_dir=cache_dir,\n",
+    "                  ipus_per_replica=1,\n",
+    "                  matmul_proportion=0.1,\n",
+    "                  explicit_ir_inference=True,\n",
+    "                 )),\n",
+    "    \n",
+    "    \"medium\": (\"openai/whisper-medium.en\",\n",
+    "        IPUConfig(executable_cache_dir=cache_dir,\n",
+    "                  ipus_per_replica=4,\n",
```
Comment (katalinic-gc) [inline notebooks/whisper-quantized-example.ipynb:187]: Not 2?

File: notebooks/whisper-quantized-example.ipynb [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+    "\n",
+    "\n",
+    "[🤗 Optimum Graphcore](https://github.com/huggingface/optimum-graphcore) is the interface between the [🤗 Transformers library](https://huggingface.co/docs/transformers/index) and [Graphcore IPUs](https://www.graphcore.ai/products/ipu).\n",
+    "It provides a set of tools enabling model parallelization and loading on IPUs, training and fine-tuning on all the tasks already supported by 🤗 Transformers while being compatible with the 🤗 Hub and every model available on it out of the box.\n",
+    "\n",
+    "> **Hardware requirements:** All the Whisper models from `whisper-tiny` to `whisper-large-v2` can run in inference mode on smallest IPU-POD4 machine.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:21]:
```suggestion
    "> **Hardware requirements:** All the Whisper models from `whisper-tiny` to `whisper-large-v2` can run in inference mode on the smallest IPU-POD4 machine.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c7a7484f",
+   "metadata": {},
+   "source": [
+    "This notebook demonstrates how to run all sizes of Whisper. All sizes will fit on an IPU-POD4:\n",
+    "\n",
+    "- `whisper-tiny`, `base` and `small` only require 1 IPU\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:134]:
```suggestion
    "- `whisper-tiny`, `whisper-base` and `whisper-small` only require 1 IPU\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "734d8d54",
+   "metadata": {},
+   "source": [
+    "The Whisper model is available on Hugging Face in several sizes, from `whisper-tiny` with 39M parameters to `whisper-large` with 1550M parameters.\n",
+    "\n",
+    "The [Whisper architecture](https://openai.com/research/whisper) is an encoder-decoder Transformer, with the audio split into 30-second chunks.\n",
+    "- For `whisper-tiny`, `small` and `base`, both encoder and decoder fit on 1 IPU.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:147]:
```suggestion
    "- For `whisper-tiny`, `whisper-small` and `whisper-base`, the encoder and decoder both fit on 1 IPU.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+    "- For `whisper-medium `, one IPU is used to place the encoder part and two others for the decoder part.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:148]:
```suggestion
    "- For `whisper-medium `, one IPU is used for the encoder and two more IPUs are used for the decoder.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+    "- For `whisper-large `, two IPUs are used to place the encoder part and two others for the decoder part.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:149]:
```suggestion
    "- For `whisper-large `, two IPUs are used for the encoder and two more IPUs are used for the decoder.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+    "## Next Steps\n",
+    "\n",
+    "The `whisper-tiny` model used here is very fast for inference and so cheap to run, but its accuracy can be improved.\n",
+    "The `whisper-base`, `whisper-small` and `whisper-medium` models have 74M, 244M and 769 M parameters respectively (compared to just 39M for `whisper-tiny`). You can try out `whisper-base`, `whisper-small` and `whisper-medium` by changing `select_whisper_config(\"small\")` (at the beginning of this notebook) to:\n",
+    "- `select_whisper_config(\"base\")`\n",
+    "- `select_whisper_config(\"small\")`\n",
+    "- `select_whisper_config(\"medium\")` respectively.\n",
+    "\n",
+    "Larger models and multilingual models are also available.\n",
+    "To access the multilingual models, remove the `.en` from the checkpoint name. Note however that the multilingual models are slightly less accurate for this English transcription task but they can be used for transcribing other languages or for translating to English. The largest model `whisper-large` has 1550M parameters and requires a 4-IPUs pipeline.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:397]:
```suggestion
    "To access the multilingual models, remove the `.en` from the checkpoint name. Note however that the multilingual models are slightly less accurate for this English transcription task but they can be used for transcribing other languages or for translating to English. The largest model `whisper-large` has 1550M parameters and requires a 4-IPU pipeline.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+    "You can try it by setting `select_whisper_config(\"large\")`\n",
+    "\n",
+    "You can also try using beam search by setting the `num_beams` parameter in the calls `parallelize` and `generate` above. `whisper-small` will fit on 1 IPU with `num_beams=5`.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:400]:
```suggestion
    "You can also try using beam search by setting `num_beams>1` in the calls to `parallelize` and `generate` above. `whisper-small` will fit on 1 IPU with `num_beams=5`.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+    "\n",
+    "For `whisper-medium` with `num_beams>1` the model will need 4 IPUs to fit. For `whisper-large` with `num_beams>1` you will need more than the IPU-POD4. On Paperspace, this is available using either an IPU-POD16 or a Bow Pod16 machine. Please contact Graphcore if you need assistance running these larger models.\n"
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:402]:
```suggestion
    "For `whisper-medium` with `num_beams>1` the model will need 4 IPUs to fit. For `whisper-large` with `num_beams>1` you will need more than the 4 IPUS in an IPU-POD4. On Paperspace, you can use either an IPU-POD16 or a Bow Pod16 machine, each with 16 IPUs. Please contact Graphcore if you need assistance running these larger models.\n"
```

File: notebooks/whisper-quantized-example.ipynb [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+    "- `whisper-medium` requires 2 IPUs\n",
+    "- `whisper-large` requires 4 IPUs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "734d8d54",
+   "metadata": {},
+   "source": [
+    "The Whisper model is available on Hugging Face in several sizes, from `whisper-tiny` with 39M parameters to `whisper-large` with 1550M parameters.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:144]:
```suggestion
    "The Whisper model is available on 🤗 Models in several sizes, from `whisper-tiny` with 39M parameters to `whisper-large` with 1550M parameters.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+   "source": [
+    "model_checkpoint, ipu_config = select_whisper_config(\"small\", custom_checkpoint=None) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eff8bcf6",
+   "metadata": {},
+   "source": [
+    "You can also use a custom checkpoint from Hugging Face Hub using the argument `custom_checkpoint` above. In this case, you have to make sure that `size` matches the checkpoint model size.\n",
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:236]:
```suggestion
    "You can also use a custom checkpoint from 🤗 Hub using the argument `custom_checkpoint` above. In this case, you have to make sure that `size` matches the checkpoint model size.\n",
```

File: notebooks/whisper-quantized-example.ipynb [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "21ff7629",
+   "metadata": {},
+   "source": [
+    "## Conclusion\n",
+    "\n",
+    "In this notebook, we demonstrated using Whisper and Group Quantization for speech recognition and transcription on the IPU.\n",
+    "We used the Optimum Graphcore package to interface between the IPU and the Hugging Face Transformers library. This meant that only a few lines of code were needed to get this state-of-the-art automated speech recognition model running on IPUs."
```
Comment (jayniep-gc) [inline notebooks/whisper-quantized-example.ipynb:413]:
```suggestion
    "We used the Optimum Graphcore package to interface between the IPU and the 🤗 Transformers library. This meant that only a few lines of code were needed to get this state-of-the-art automated speech recognition model running on IPUs."
```

**Review by jayniep-gc — COMMENTED** [review]
Some fixes from a Tech Docs perspective

File: notebooks/whisper-quantized-example.ipynb (hunk 3 above)
Comment (jimypbr) [reply notebooks/whisper-quantized-example.ipynb:182]: explicit IR is needed when using beam search iirc. The notebook invites users to try beam search at the end. Best leave it in I think. Same with the matmul proportion. Alternatively we could just add another section that does beam search and put the different IPUConfigs in there.

Comment (jimypbr) [reply notebooks/whisper-quantized-example.ipynb:189]: 0.1 wasn't necessary in the end. removed.

File: notebooks/whisper-quantized-example.ipynb (hunk 2 above)
Comment (jimypbr) [reply notebooks/whisper-quantized-example.ipynb:97]: We can come back to this with the release.

**Review by katalinic-gc — APPROVED** [review]

**Pull request merged**